src/registry.rs

```diff
@@ -0,0 +1,135 @@
+use self::code_from_cargo::Kind;
+use crate::errors::*;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::env::current_dir;
+use std::path::{Path, PathBuf};
+use url::Url;
+
+const CRATES_IO_INDEX: &str = "https://github.com/rust-lang/crates.io-index";
+const CRATES_IO_REGISTRY: &str = "crates-io";
+
+pub fn registry_path() -> Result<PathBuf> {
+    Ok(cargo_home()?
+        .join("registry")
+        .join("index")
+        .join(short_name(&registry_url()?)))
+}
+
+#[derive(Debug, Deserialize)]
+struct Source {
+    #[serde(rename = "replace-with")]
+    replace_with: Option<String>,
+    registry: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct CargoConfig {
+    #[serde(default)]
+    sources: HashMap<String, Source>,
+}
+
+fn cargo_home() -> Result<PathBuf> {
+    let default_cargo_home = dirs::home_dir()
+        .map(|x| x.join(".cargo"))
+        .chain_err(|| ErrorKind::ReadHomeDirFailure)?;
+    let cargo_home = std::env::var("CARGO_HOME")
+        .map(PathBuf::from)
+        .unwrap_or(default_cargo_home);
+    Ok(cargo_home)
+}
+
+pub fn registry_url() -> Result<Url> {
+    // TODO support local registry sources, directory sources, git sources: https://doc.rust-lang.org/cargo/reference/source-replacement.html?highlight=replace-with#source-replacement
+    fn read_config(registries: &mut HashMap<String, Source>, path: impl AsRef<Path>) -> Result<()> {
+        let content = std::fs::read(path)?;
+        let config =
+            toml::from_slice::<CargoConfig>(&content).map_err(|_| ErrorKind::InvalidCargoConfig)?;
+        for (key, value) in config.sources {
+            registries.entry(key).or_insert(value);
+        }
+        Ok(())
+    }
+
+    let mut registries: HashMap<String, Source> = HashMap::new();
+
+    // ref: https://doc.rust-lang.org/cargo/reference/config.html#hierarchical-structure
+    for work_dir in current_dir()?.ancestors() {
```

Comment:
maybe instead of `current_dir` we should first try `--manifest-path` first? we could be executing this command from a different dir, also this is worth mentioning in the readme

Reply:
It seems hard to pass a manifest path to here, is that ok if I want to use global variable? I want to put those config to a global variable.

Reply:
hmm, I'm not sure why is it hard to add an argument to registry_url (and all functions up the stack)?

Reply:
done.